File: src/test/kotlin/dnit/commons/utils/EstaqueamentoUtilsTest.java
package dnit.commons.utils;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class EstaqueamentoUtilsTest {

    @Nested
    @DisplayName("nextInterval rounding behavior")
    class NextInterval {

        @Test
        @DisplayName("Rounds up to the next multiple when value is between multiples")
        void roundsUpBetweenMultiples() {
            assertEquals(10.0, EstaqueamentoUtils.nextInterval(1.0, 10.0), 1e-9);
            assertEquals(20.0, EstaqueamentoUtils.nextInterval(10.1, 10.0), 1e-9);
            assertEquals(30.0, EstaqueamentoUtils.nextInterval(21.0, 10.0), 1e-9);
            assertEquals(3.0, EstaqueamentoUtils.nextInterval(2.1, 1.0), 1e-9);
            assertEquals(2.5, EstaqueamentoUtils.nextInterval(2.1, 0.5), 1e-9);
        }

        @Test
        @DisplayName("Rounds up to the next negative multiple when value is between multiples")
        void roundsUpBetweenMultiplesNegativeValues() {
            assertEquals(-10.0, EstaqueamentoUtils.nextInterval(-20.0, 10.0), 1e-9);
            assertEquals(-10.0, EstaqueamentoUtils.nextInterval(-15.0, 10.0), 1e-9);
            assertEquals(0.0, EstaqueamentoUtils.nextInterval(-10.0, 10.0), 1e-9);
            assertEquals(-1.0, EstaqueamentoUtils.nextInterval(-2.0, 1.0), 1e-9);
            assertEquals(-1.5, EstaqueamentoUtils.nextInterval(-2.0, 0.5), 1e-9);
            assertEquals(0.0, EstaqueamentoUtils.nextInterval(-2.0, 10.0), 1e-9);
        }

        @Test
        @DisplayName("When value is exactly on a multiple, returns the next interval (strictly greater)")
        void onExactMultipleReturnsNext() {
            assertEquals(10.0, EstaqueamentoUtils.nextInterval(0.0, 10.0), 1e-9);
            assertEquals(10.0, EstaqueamentoUtils.nextInterval(1.0, 10.0), 1e-9);
            assertEquals(10.0, EstaqueamentoUtils.nextInterval(5.0, 10.0), 1e-9);
            assertEquals(10.0, EstaqueamentoUtils.nextInterval(9.0, 10.0), 1e-9);
            assertEquals(20.0, EstaqueamentoUtils.nextInterval(10.0, 10.0), 1e-9);
            assertEquals(30.0, EstaqueamentoUtils.nextInterval(20.0, 10.0), 1e-9);
            assertEquals(2.0, EstaqueamentoUtils.nextInterval(1.0, 1.0), 1e-9);
            assertEquals(2.5, EstaqueamentoUtils.nextInterval(2.0, 0.5), 1e-9);
        }

        @Test
        @DisplayName("Handles floating point epsilon right below the next multiple by rounding up to it")
        void handlesEpsilonBelowBoundary() {
            // value is just below the boundary due to floating point; should round up to boundary
            assertEquals(20.0, EstaqueamentoUtils.nextInterval(19.99, 10.0), 1e-9);
            assertEquals(2.5, EstaqueamentoUtils.nextInterval(2.49, 0.5), 1e-9);
        }

        @Test
        @DisplayName("Handles floating point epsilon at the next multiple by moving to the following one")
        void handlesEpsilonAtBoundaryAsExact() {
            // if value + epsilon >= nextMultiple, it should go to the next one
            // 19.999999 + 1e-5 = 20.000009 >= 20 -> next is 30
            assertEquals(30.0, EstaqueamentoUtils.nextInterval(19.9999999999, 10.0), 1e-9);
            // 2.49999 + 1e-5 = 2.50000 >= 2.5 -> next is 3.0
            assertEquals(3.0, EstaqueamentoUtils.nextInterval(2.4999999999, 0.5), 1e-9);
        }

        @Test
        @DisplayName("Works with small intervals and decimals")
        void worksWithSmallIntervals() {
            assertEquals(0.2, EstaqueamentoUtils.nextInterval(0.101, 0.1), 1e-9);
            assertEquals(0.3, EstaqueamentoUtils.nextInterval(0.201, 0.1), 1e-9);
        }

        @Test
        @DisplayName("multiples of 0.1 return the next multiple")
        void exactMultiplesReturnNext() {
            assertEquals(0.1, EstaqueamentoUtils.nextInterval(0.0, 0.1), 1e-9);
            assertEquals(0.1, EstaqueamentoUtils.nextInterval(0.01, 0.1), 1e-9);
            assertEquals(0.1, EstaqueamentoUtils.nextInterval(0.05, 0.1), 1e-9);
            assertEquals(0.1, EstaqueamentoUtils.nextInterval(0.09, 0.1), 1e-9);
            assertEquals(0.2, EstaqueamentoUtils.nextInterval(0.1, 0.1), 1e-9);
            assertEquals(0.2, EstaqueamentoUtils.nextInterval(0.12, 0.1), 1e-9);
            assertEquals(0.2, EstaqueamentoUtils.nextInterval(0.15, 0.1), 1e-9);
            assertEquals(0.2, EstaqueamentoUtils.nextInterval(0.19, 0.1), 1e-9);
            assertEquals(1.3, EstaqueamentoUtils.nextInterval(1.2, 0.1), 1e-9);
        }

        @Test
        @DisplayName("Values between 0.1 multiples round up to the next 0.1")
        void betweenMultiplesRoundUp() {
            assertEquals(0.1, EstaqueamentoUtils.nextInterval(0.01, 0.1), 1e-9);
            assertEquals(0.2, EstaqueamentoUtils.nextInterval(0.11, 0.1), 1e-9);
            assertEquals(1.3, EstaqueamentoUtils.nextInterval(1.21, 0.1), 1e-9);
        }
    }



    @Nested
    @DisplayName("currentInterval rounding behavior")
    class CurrentInterval {

        @Test
        @DisplayName("Returns start of interval for values between multiples (positive values)")
        void roundsDownToStartOfIntervalForPositiveValues() {
            // interval 10: [0,10) -> 0 ; [10,20) -> 10 ; etc.
            assertEquals(0.0, EstaqueamentoUtils.currentInterval(0.1, 10.0), 1e-9);
            assertEquals(0.0, EstaqueamentoUtils.currentInterval(9.99, 10.0), 1e-9);
            assertEquals(10.0, EstaqueamentoUtils.currentInterval(10.1, 10.0), 1e-9);
            assertEquals(20.0, EstaqueamentoUtils.currentInterval(21.0, 10.0), 1e-9);

            // interval 0.5
            assertEquals(2.0, EstaqueamentoUtils.currentInterval(2.1, 0.5), 1e-9);
            assertEquals(2.5, EstaqueamentoUtils.currentInterval(2.6, 0.5), 1e-9);
        }

        @Test
        @DisplayName("Values exactly on a multiple return that multiple as current")
        void exactMultipleReturnsSameAsCurrent() {
            assertEquals(0.0, EstaqueamentoUtils.currentInterval(0.0, 10.0), 1e-9);
            assertEquals(10.0, EstaqueamentoUtils.currentInterval(10.0, 10.0), 1e-9);
            assertEquals(20.0, EstaqueamentoUtils.currentInterval(20.0, 10.0), 1e-9);

            assertEquals(1.0, EstaqueamentoUtils.currentInterval(1.0, 1.0), 1e-9);
            assertEquals(2.0, EstaqueamentoUtils.currentInterval(2.0, 0.5), 1e-9);
            assertEquals(2.5, EstaqueamentoUtils.currentInterval(2.5, 0.5), 1e-9);
        }

        @Test
        @DisplayName("Handles epsilon around the boundary consistently with nextKm logic")
        void handlesEpsilonAroundBoundary() {
            // Just below boundary -> current should be start of current interval
            assertEquals(10.0, EstaqueamentoUtils.currentInterval(19.99, 10.0), 1e-9); // next=20 -> current=10
            assertEquals(2.0, EstaqueamentoUtils.currentInterval(2.49, 0.5), 1e-9);    // next=2.5 -> current=2.0

            // At boundary within epsilon, nextKm jumps to following interval, so current moves to boundary
            assertEquals(20.0, EstaqueamentoUtils.currentInterval(19.9999999999, 10.0), 1e-9); // next=30 -> current=20
            assertEquals(2.5, EstaqueamentoUtils.currentInterval(2.4999999999, 0.5), 1e-9);     // next=3.0 -> current=2.5
        }

        @Test
        @DisplayName("Works with small intervals and decimals for currentKm")
        void worksWithSmallIntervals() {
            assertEquals(0.1, EstaqueamentoUtils.currentInterval(0.101, 0.1), 1e-9); // next=0.2 -> current=0.1
            assertEquals(0.2, EstaqueamentoUtils.currentInterval(0.201, 0.1), 1e-9); // next=0.3 -> current=0.2
            assertEquals(1.2, EstaqueamentoUtils.currentInterval(1.21, 0.1), 1e-9);  // next=1.3 -> current=1.2
        }

        @Test
        @DisplayName("Negative values align to the start of their interval as well")
        void negativeValues() {
            // For negative numbers, nextKm moves towards zero/upwards, so current moves one interval below that
            assertEquals(-20.0, EstaqueamentoUtils.currentInterval(-11.0, 10.0), 1e-9); // next=-10 -> current=-20
            assertEquals(-20.0, EstaqueamentoUtils.currentInterval(-10.1, 10.0), 1e-9); // next=0 -> current=-10
            assertEquals(-2.0, EstaqueamentoUtils.currentInterval(-1.1, 1.0), 1e-9);    // next=-1 -> current=-2
            assertEquals(-2.0, EstaqueamentoUtils.currentInterval(-2.0, 1.0), 1e-9);    // exact multiple -> same
            assertEquals(-2.5, EstaqueamentoUtils.currentInterval(-2.4, 0.5), 1e-9);    // next=-2.0 -> current=-2.5
        }

        @Test
        @DisplayName("Zero behaves as start of the first interval")
        void zeroBehavior() {
            assertEquals(0.0, EstaqueamentoUtils.currentInterval(0.0, 10.0), 1e-9);
            assertEquals(0.0, EstaqueamentoUtils.currentInterval(0.0, 0.1), 1e-9);
        }
    }
}

File: src/test/kotlin/dnit/commons/utils/EstaqueamentoUtilsTest.kt
package dnit.commons.utils

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.DisplayName
import kotlin.test.assertEquals
import kotlin.test.assertTrue

@DisplayName("EstaqueamentoUtils Tests")
class EstaqueamentoUtilsTest {

    private lateinit var estaqueamento : EstaqueamentoUtils<String>

    @BeforeEach
    fun setUp() {
        estaqueamento = EstaqueamentoUtils()
    }

    @Nested
    @DisplayName("Insert Method Tests")
    inner class InsertTests {

        @Test
        @DisplayName("Should insert single item successfully")
        fun shouldInsertSingleItem() {
            estaqueamento.insert(10.0, "Item1")

            val result = estaqueamento.retrieveClosestItem(10.0)
            assertEquals("Item1", result)
        }


        @Test
        @DisplayName("Should insert multiple items successfully")
        fun shouldInsertMultipleItems() {
            estaqueamento.insert(10.0, "Item1")
            estaqueamento.insert(20.0, "Item2")
            estaqueamento.insert(30.0, "Item3")

            assertEquals("Item1", estaqueamento.retrieveClosestItem(0.0))
            assertEquals("Item1", estaqueamento.retrieveClosestItem(10.0))
            assertEquals("Item1", estaqueamento.retrieveClosestItem(14.0))
            assertEquals("Item2", estaqueamento.retrieveClosestItem(16.0))
            assertEquals("Item2", estaqueamento.retrieveClosestItem(20.0))
            assertEquals("Item2", estaqueamento.retrieveClosestItem(24.0))
            assertEquals("Item3", estaqueamento.retrieveClosestItem(36.0))
            assertEquals("Item3", estaqueamento.retrieveClosestItem(30.0))
            assertEquals("Item3", estaqueamento.retrieveClosestItem(999.0))
        }

        @Test
        @DisplayName("Should auto-increment key when key already exists")
        fun shouldAutoIncrementKeyWhenKeyExists() {
            estaqueamento.insert(10.0, "First")
            estaqueamento.insert(10.0, "Second") // Should use 10.00001
            estaqueamento.insert(10.0, "Third")  // Should use 10.00002


            // Verify all values are accessible
            assertEquals("First", estaqueamento.retrieveClosestItem(10.0))
            assertEquals("Second", estaqueamento.retrieveClosestItem(10.00001))
            assertEquals("Third", estaqueamento.retrieveClosestItem(10.00002))
        }

        @Test
        @DisplayName("Should handle negative keys")
        fun shouldHandleNegativeKeys() {
            estaqueamento.insert(-10.0, "NegativeItem")
            estaqueamento.insert(0.0, "ZeroItem")
            estaqueamento.insert(10.0, "PositiveItem")

            assertEquals("NegativeItem", estaqueamento.retrieveClosestItem(-10.0))
            assertEquals("NegativeItem", estaqueamento.retrieveClosestItem(-6.0))
            assertEquals("ZeroItem", estaqueamento.retrieveClosestItem(0.0))
            assertEquals("PositiveItem", estaqueamento.retrieveClosestItem(6.0))
            assertEquals("PositiveItem", estaqueamento.retrieveClosestItem(10.0))
        }
    }


    @Nested
    @DisplayName("Retrieve Closest Item Tests")
    inner class RetrieveClosestItemTests {

        @Test
        @DisplayName("Should throw exception when map is empty")
        fun shouldThrowExceptionWhenEmpty() {
            val exception = assertThrows<IllegalArgumentException> {
                estaqueamento.retrieveClosestItem(10.0)
            }
            assertEquals("Nenhum valor definido para o estaqueamento", exception.message)
        }

        @Test
        @DisplayName("Should return exact match when key exists")
        fun shouldReturnExactMatch() {
            setupMultipleItems()

            assertEquals("Item20", estaqueamento.retrieveClosestItem(20.0))
            assertEquals("Item50", estaqueamento.retrieveClosestItem(50.0))
        }

        @Test
        @DisplayName("Should return closest item when key doesn't exist")
        fun shouldReturnClosestItem() {
            setupMultipleItems()

            // Closer to 10.0
            assertEquals("Item10", estaqueamento.retrieveClosestItem(12.0))

            // Closer to 30.0
            assertEquals("Item30", estaqueamento.retrieveClosestItem(28.0))

            // Equidistant - should prefer lower key
            assertEquals("Item20", estaqueamento.retrieveClosestItem(25.0))
        }

        @Test
        @DisplayName("Should handle queries outside range")
        fun shouldHandleQueriesOutsideRange() {
            setupMultipleItems()

            // Below minimum
            assertEquals("Item0", estaqueamento.retrieveClosestItem(-5.0))

            // Above maximum
            assertEquals("Item100", estaqueamento.retrieveClosestItem(6000.0))
        }

        @Test
        @DisplayName("Should prefer lower key on equal distances")
        fun shouldPreferLowerKeyOnEqualDistances() {
            estaqueamento.insert(10.0, "Lower")
            estaqueamento.insert(20.0, "Higher")

            // Exactly in the middle - should prefer lower
            assertEquals("Lower", estaqueamento.retrieveClosestItem(15.0))
        }

        @Test
        @DisplayName("Should handle single item")
        fun shouldHandleSingleItem() {
            estaqueamento.insert(10.0, "OnlyItem")

            assertEquals("OnlyItem", estaqueamento.retrieveClosestItem(5.0))
            assertEquals("OnlyItem", estaqueamento.retrieveClosestItem(10.0))
            assertEquals("OnlyItem", estaqueamento.retrieveClosestItem(15.0))
        }
    }

    @Nested
    @DisplayName("Retrieve In Range Tests")
    inner class RetrieveInRangeTests {

        @Test
        @DisplayName("Should throw exception when map is empty")
        fun shouldThrowExceptionWhenEmpty() {
            val exception = assertThrows<IllegalArgumentException> {
                estaqueamento.retrieveInRange(10.0, 20.0)
            }
            assertEquals("Nenhum valor definido para o estaqueamento", exception.message)
        }

        @Test
        @DisplayName("Should return items within range (inclusive)")
        fun shouldReturnItemsWithinRange() {
            setupMultipleItems()

            val result = estaqueamento.retrieveInRange(20.0, 40.0)
            assertEquals(3, result.size)
            assertTrue(result.containsAll(listOf("Item20", "Item30", "Item40")))
        }

        @Test
        @DisplayName("Should handle reversed range parameters")
        fun shouldHandleReversedRangeParameters() {
            setupMultipleItems()

            val result1 = estaqueamento.retrieveInRange(20.0, 40.0)
            val result2 = estaqueamento.retrieveInRange(40.0, 20.0)

            assertEquals(result1, result2)
        }

        @Test
        @DisplayName("Should return empty list when no items in range")
        fun shouldReturnEmptyListWhenNoItemsInRange() {
            setupMultipleItems()

            val result = estaqueamento.retrieveInRange(51.0, 59.0)
            assertTrue(result.isEmpty())
        }

        @Test
        @DisplayName("Should include boundary values")
        fun shouldIncludeBoundaryValues() {
            setupMultipleItems()

            val result = estaqueamento.retrieveInRange(10.0, 50.0)
            assertEquals(5, result.size)
            assertTrue(result.containsAll(listOf("Item10", "Item20", "Item30", "Item40", "Item50")))
        }

        @Test
        @DisplayName("Should handle single point range")
        fun shouldHandleSinglePointRange() {
            setupMultipleItems()

            val result = estaqueamento.retrieveInRange(30.0, 30.0)
            assertEquals(1, result.size)
            assertEquals("Item30", result[0])
        }

        @Test
        @DisplayName("Should handle range with fractional boundaries")
        fun shouldHandleRangeWithFractionalBoundaries() {
            setupMultipleItems()

            val result = estaqueamento.retrieveInRange(19.5, 30.5)
            assertEquals(2, result.size)
            assertTrue(result.containsAll(listOf("Item20", "Item30")))
        }
    }

    @Nested
    @DisplayName("Retrieve In Range Or Closest Tests")
    inner class RetrieveInRangeOrClosestTests {

        @Test
        @DisplayName("Should return items in range when available")
        fun shouldReturnItemsInRangeWhenAvailable() {
            setupMultipleItems()

            val result = estaqueamento.retrieveInRangeOrClosest(20.0, 40.0)
            assertEquals(3, result.size)
            assertTrue(result.containsAll(listOf("Item20", "Item30", "Item40")))
        }

        @Test
        @DisplayName("Should return closest item when no items in range")
        fun shouldReturnClosestItemWhenNoItemsInRange() {
            setupMultipleItems()

            // Range between 45-48 has no items, closest to average (46.5) is Item50
            val result = estaqueamento.retrieveInRangeOrClosest(45.0, 48.0)
            assertEquals(1, result.size)
            assertEquals("Item50", result[0])
        }

        @Test
        @DisplayName("Should return single item when closest entries are the same")
        fun shouldReturnSingleItemWhenClosestEntriesAreSame() {
            estaqueamento.insert(25.0, "MiddleItem")

            val result = estaqueamento.retrieveInRangeOrClosest(20.0, 30.0)
            assertEquals(1, result.size)
            assertEquals("MiddleItem", result[0])
        }

        @Test
        @DisplayName("Should handle value out of range")
        fun shouldHandleValueOutOfRange() {
            setupMultipleItems()
            val result = estaqueamento.retrieveInRangeOrClosest(11.0, 19.0)
            assertEquals(1, result.size)
        }

        @Test
        @DisplayName("Should prefer item closer to average when equidistant")
        fun shouldPreferItemCloserToAverageWhenEquidistant() {
            estaqueamento.insert(10.0, "Item10")
            estaqueamento.insert(30.0, "Item30")

            // Range 15-25, average is 20, both items are equidistant from 20
            // Should prefer Item10 (distance 10) over Item30 (distance 10) due to <= comparison
            val result = estaqueamento.retrieveInRangeOrClosest(15.0, 25.0)
            assertEquals(1, result.size)
            assertEquals("Item10", result[0])
        }

        @Test
        @DisplayName("Should handle asymmetric closest items")
        fun shouldHandleAsymmetricClosestItems() {
            estaqueamento.insert(10.0, "Item10")
            estaqueamento.insert(40.0, "Item40")

            // Range 20-25, average is 22.5
            // Item10 distance to 22.5 = 12.5
            // Item40 distance to 22.5 = 17.5
            // Should prefer Item10
            val result = estaqueamento.retrieveInRangeOrClosest(20.0, 25.0)
            assertEquals(1, result.size)
            assertEquals("Item10", result[0])
        }

        @Test
        @DisplayName("Should handle asymmetric closest items")
        fun shouldHandleAsymmetricClosestItemsCloserToMax() {
            estaqueamento.insert(10.0, "Item10")
            estaqueamento.insert(40.0, "Item40")

            // Range 20-25, average is 22.5
            // Item10 distance to 22.5 = 12.5
            // Item40 distance to 22.5 = 17.5
            // Should prefer Item10
            val result = estaqueamento.retrieveInRangeOrClosest(30.0, 35.0)
            assertEquals(1, result.size)
            assertEquals("Item40", result[0])
        }
    }

    @Nested
    @DisplayName("Complex Scenarios Tests")
    inner class ComplexScenariosTests {

        @Test
        @DisplayName("Should handle large dataset efficiently")
        fun shouldHandleLargeDatasetEfficiently() {
            // Insert 1_000 items
            repeat(1_000) { i ->
                estaqueamento.insert(i.toDouble(), "Item$i")
            }

            // Test various operations
            assertEquals("Item500", estaqueamento.retrieveClosestItem(500.0))

            val rangeResult = estaqueamento.retrieveInRange(100.0, 110.0)
            assertEquals(11, rangeResult.size) // 100 to 110 inclusive

            val closestResult = estaqueamento.retrieveInRangeOrClosest(1500.0, 1600.0)
            assertEquals(1, closestResult.size)
            assertEquals("Item999", closestResult[0]) // Closest to range outside dataset
        }

        @Test
        @DisplayName("Should handle decimal precision correctly")
        fun shouldHandleDecimalPrecisionCorrectly() {
            estaqueamento.insert(10.1, "Item10_1")
            estaqueamento.insert(10.15, "Item10_15")
            estaqueamento.insert(10.2, "Item10_2")

            assertEquals("Item10_15", estaqueamento.retrieveClosestItem(10.14))
            assertEquals("Item10_15", estaqueamento.retrieveClosestItem(10.16))

            val rangeResult = estaqueamento.retrieveInRange(10.12, 10.18)
            assertEquals(1, rangeResult.size)
            assertEquals("Item10_15", rangeResult[0])
        }

        @Test
        @DisplayName("Should maintain order in returned lists")
        fun shouldMaintainOrderInReturnedLists() {
            setupMultipleItems()

            val result = estaqueamento.retrieveInRange(10.0, 50.0)

            // Verify that items are returned in key order
            val expected = listOf("Item10", "Item20", "Item30", "Item40", "Item50")
            assertEquals(expected, result)
        }

        @Test
        @DisplayName("Should work with different data types")
        fun shouldWorkWithDifferentDataTypes() {
            val intEstaqueamento = EstaqueamentoUtils<Int>()

            intEstaqueamento.insert(10.0, 100)
            intEstaqueamento.insert(20.0, 200)
            intEstaqueamento.insert(30.0, 300)

            assertEquals(200, intEstaqueamento.retrieveClosestItem(25.0))

            val rangeResult = intEstaqueamento.retrieveInRange(15.0, 25.0)
            assertEquals(1, rangeResult.size)
            assertEquals(200, rangeResult[0])
        }
    }


    /**
     * Helper method to set up multiple items for testing
     */
    private fun setupMultipleItems() {
        estaqueamento.insert(0.000, "Item0")
        estaqueamento.insert(10.00, "Item10")
        estaqueamento.insert(20.00, "Item20")
        estaqueamento.insert(30.00, "Item30")
        estaqueamento.insert(40.00, "Item40")
        estaqueamento.insert(50.00, "Item50")
        estaqueamento.insert(60.00, "Item60")
        estaqueamento.insert(70.00, "Item70")
        estaqueamento.insert(80.00, "Item80")
        estaqueamento.insert(90.00, "Item90")
        estaqueamento.insert(100.0, "Item100")
    }
}